src/lib/structures/objective.ml

```diff
@@ -0,0 +1,118 @@
+(*                                                                        *)
+(*     Alt-Ergo: The SMT Solver For Software Verification                 *)
+(*     Copyright (C) 2013-2023 --- OCamlPro SAS                           *)
+(*                                                                        *)
+(*     This file is distributed under the terms of OCamlPro               *)
+(*     Non-Commercial Purpose License, version 1.                         *)
+(*                                                                        *)
+(*     As an exception, Alt-Ergo Club members at the Gold level can       *)
+(*     use this file under the terms of the Apache Software License       *)
+(*     version 2.0.                                                       *)
+(*                                                                        *)
+(*     ---------------------------------------------------------------    *)
+(*                                                                        *)
+(*     The Alt-Ergo theorem prover                                        *)
+(*                                                                        *)
+(*     Sylvain Conchon, Evelyne Contejean, Francois Bobot                 *)
+(*     Mohamed Iguernelala, Stephane Lescuyer, Alain Mebsout              *)
+(*                                                                        *)
+(*     CNRS - INRIA - Universite Paris Sud                                *)
+(*                                                                        *)
+(*     Until 2013, some parts of this code were released under            *)
+(*     the Apache Software License version 2.0.                           *)
+(*                                                                        *)
+(*     ---------------------------------------------------------------    *)
+(*                                                                        *)
+(*     More details can be found in the directory licenses/               *)
+(*                                                                        *)
+(**************************************************************************)
+
+module Function = struct
+  type t = {
+    e : Expr.t;
+    to_max : bool;
+  }
+
+  let equal { e = e1; _ } { e = e2; _ } = Expr.equal e1 e2
+
+  let mk ~to_max e = { e; to_max }
+
+  let pp ppf { e; _ } = Expr.print ppf e
+end
+
+module Value = struct
+  type limit_kind =
+    | Above
+    | Below
+
+  type t =
+    | Minfinity
+    | Pinfinity
+    | Value of Expr.t
+    | Limit of limit_kind * Expr.t
+    | Unknown
+
+  let pp ppf v =
+    match v with
+    | Minfinity -> Fmt.pf ppf "-oo"
+    | Pinfinity -> Fmt.pf ppf "+oo"
+    | Value w -> Expr.print ppf w
+    | Limit (Above, w) -> Fmt.pf ppf "(+ %a epsilon)" Expr.print w
+    | Limit (Below, w) -> Fmt.pf ppf "(- %a epsilon)" Expr.print w
+    | Unknown -> Fmt.pf ppf "(interval (- oo) (+ oo))"
```

Comment:
Why the space between `-` (`+`) and `oo` ?

Reply:
No good reason. We can consider `oo` as a constant and apply an unary `+` or an unary `-` on it. Or we consider two constants `+oo`, `-oo`. My code is not consistent with that. This is the output of `arith3.optimize.smt2` by `z3`:
```
sat
(
  (define-fun x () Int
    0)
  (define-fun z () Int
    0)
  (define-fun y () Int
    0)
)
(objectives
 (x oo)
 (y  (interval (* (- 1) oo) oo))
 (z  (interval (* (- 1) oo) oo))
)
```
So we can follow the same principal and consider `oo` as a constant and apply usual arithmetical operator on it.

Notice that mathematically, it's a heresy. For instance, `(- oo oo)` doesn't make any sense.

Reply:
Sure, I said we have to be consistent and we use the Z3 convention.